From 9eff5cf52c3e2841fdf03db58412cac32c3db0b2 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@notvalid.com>
Date: Wed, 14 Oct 2026 16:31:13 -0700
Subject: [PATCH] Adapt Dogecoin activation boundaries and header hashing to
 current ABC

Keep current height/time APIs without narrowing disabled upgrade times. Disable newer hard-consensus transaction version checks on Dogecoin mainnet, remove incompatible eCash snapshot commitments, and preserve optional historical RTT override interfaces.

diff --git a/src/consensus/params.h b/src/consensus/params.h
index 1b7f99b4868b7caf5a6f4d9e71c208a9fc702711..71056c718058b4d49199f4423c33887533a02727 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -73,9 +73,11 @@ struct Params {
      * Unix time used for MTP activation of 15 November 2026 12:00:00 UTC
      * upgrade
      */
-    int mengerActivationTime;
+    int64_t rttActivationTime = 0x7ffffffe;
+    int64_t rttOneBlockActivationTime = 0x7ffffffe;
+    int64_t mengerActivationTime;
     /** Unix time used for MTP activation of 15 May 2027 12:00:00 UTC upgrade */
-    int sandersActivationTime;
+    int64_t sandersActivationTime;
 
     /** Dogecoin "Digishield" activation height */
     int digishieldHeight;
diff --git a/src/init.cpp b/src/init.cpp
index 0b2c28eb82a3d2338db7ce875e1a2b1f5c84f4a8..69d3240bc755f0b0ecc79d72833b49e7933388a2 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1678,6 +1678,7 @@ void SetupServerArgs(NodeContext &node) {
 
     // Add the hidden options
     argsman.AddHiddenArgs(hidden_args);
+    argsman.AddHiddenArgs({"-augustoactivationtime", "-shibusawaactivationtime"});
 }
 
 static bool fHaveGenesis = false;
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index c6991ae3151705eeafe603f1cebac0155384f8cd..d74c0e42db77f2fbf008b44f07dc6303e3b3a44f 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -128,7 +128,7 @@ public:
         consensus.gravitonHeight = 0x7ffffffe;
         consensus.phononHeight = 0x7ffffffe;
         consensus.axionHeight = 0x7ffffffe;
-        consensus.wellingtonHeight = 800000;    // keep alive for tests
+        consensus.wellingtonHeight = 0x7ffffffe;    // keep alive for tests
         consensus.cowperthwaiteHeight = 900000; // keep alive for tests
         consensus.mengerActivationTime = std::numeric_limits<int64_t>::max();
         consensus.sandersActivationTime = std::numeric_limits<int64_t>::max();
@@ -196,26 +196,8 @@ public:
 
         checkpointData = CheckpointData(ChainType::MAIN);
 
-        m_assumeutxo_data = {
-            // v0.31.5
-            {.height = 896'800,
-             .hash_serialized =
-                 AssumeutxoHash{uint256S("0x2f783c045f353b4a900139f8a66c7bcbb62"
-                                         "c93a4f298472d77dad9bfb0186665")},
-             .m_chain_tx_count = 299'407'257,
-             .blockhash =
-                 BlockHash{uint256S("0x0000000000000000297efb200794348b44bff4bf"
-                                    "b31716cf64dc45bac0a251ea")}},
-            // v0.32.0
-            {.height = 916'000,
-             .hash_serialized =
-                 AssumeutxoHash{uint256S("0x20f077a8fcc08bb9cb3753df845c9a4257c"
-                                         "e500684dfa5a95d1061701c4fa35d")},
-             .m_chain_tx_count = 299'856'530,
-             .blockhash =
-                 BlockHash{uint256S("0x00000000000000003fc542691c35873ba4ba7a44"
-                                    "05ce612a62f121988fb8a46d")}},
-        };
+        // ABC snapshot commitments do not describe the Dogecoin chain.
+        m_assumeutxo_data = {};
 
         // Data as of block
         // 000000000000000001d2ce557406b017a928be25ee98906397d339c3f68eec5d
@@ -347,26 +329,8 @@ public:
 
         checkpointData = CheckpointData(ChainType::TESTNET);
 
-        m_assumeutxo_data = {
-            // v0.31.5
-            {.height = 1'661'000,
-             .hash_serialized =
-                 AssumeutxoHash{uint256S("0xc7a2aa5dfdbafa2d6a6613d254d25a2ab9d"
-                                         "893c01099d241f1e7a3785cb5f50f")},
-             .m_chain_tx_count = 63'977'749,
-             .blockhash =
-                 BlockHash{uint256S("0x000000000000c7d18ee9b71a1ab4d8d21aa9d758"
-                                    "7bf260e93df029ccb392d403")}},
-            // v0.32.0
-            {.height = 1'680'000,
-             .hash_serialized =
-                 AssumeutxoHash{uint256S("0x53026f5c5f3bfdbfb3acda49d5531dfcbbb"
-                                         "544cfa0e15e3fcecc22d681aa9986")},
-             .m_chain_tx_count = 63'999'716,
-             .blockhash =
-                 BlockHash{uint256S("0x000000000003c4467ce74a73c902e80b5924cfdf"
-                                    "2695bea8991963f26ac6f4b1")}},
-        };
+        // ABC snapshot commitments do not describe the Dogecoin chain.
+        m_assumeutxo_data = {};
 
         // Data as of block
         // 000000000005b07ecf85563034d13efd81c1a29e47e22b20f4fc6919d5b09cd6
@@ -447,6 +411,8 @@ public:
         consensus.cowperthwaiteHeight = 0;
 
         // Nov 15, 2026 12:00:00 UTC protocol upgrade
+        consensus.rttActivationTime = 1731672000;
+        consensus.rttOneBlockActivationTime = 1763208000;
         consensus.mengerActivationTime = 1794744000;
 
         // May 15, 2027 12:00:00 UTC protocol upgrade
diff --git a/src/primitives/baseheader.cpp b/src/primitives/baseheader.cpp
index ecaa837e1ae95274be38f732632457d244aa6779..2790d13ccf9c47568e46bdbc46a7464f1ee895f8 100644
--- a/src/primitives/baseheader.cpp
+++ b/src/primitives/baseheader.cpp
@@ -10,7 +10,9 @@
 #include <serialize.h>
 
 BlockHash CBaseBlockHeader::GetHash() const {
-    return BlockHash(SerializeHash(*this));
+    HashWriter writer;
+    writer << *this;
+    return BlockHash(writer.GetHash());
 }
 
 BlockHash CBaseBlockHeader::GetPowHash() const {
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index 93eec38094ce4a1fd56f67cf6636db6e13d39db0..1277a991a0d975cd58943f63b433dc8f5240c0a0 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -1263,7 +1263,7 @@ static RPCHelpMan getblocktemplate() {
                 UniValue rtt(UniValue::VOBJ);
 
                 UniValue prevHeaderTimes(UniValue::VARR);
-                for (size_t i : {1, 2, 5, 11, 17}) {
+                for (size_t i : GetRTTFactorIndices(consensusParams, pindexPrev)) {
                     prevHeaderTimes.push_back(prevHeaderReceivedTime[i]);
                 }
 
